### PR TITLE
[Pending testing] Palpatim/finish transcribe wss

### PR DESCRIPTION
**Pull request by palpatim**
*Issue #, if available:*

*Description of changes:*

Finishes transcribe streaming implementation
Includes addition of unit tests for SigV4 presigned URL signing

**Note**

There are a *lot* of changed files in this commit, but many of them are either unit test cases for SigV4, or Ruby scripts to generate the Swift version of those unit tests.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: AWSCore/Service/AWSService.m [hunk 1]
```diff
@@ -293,6 +293,7 @@ - (id)copyWithZone:(NSZone *)zone {
 static NSString *const AWSServiceNameKinesisVideo = @"kinesisvideo";
 static NSString *const AWSServiceNameKinesisVideoArchivedMedia = @"kinesisvideo";
 static NSString *const AWSServiceNameSageMakerRuntime = @"sagemaker";
+static NSString *const AWSServiceNameTranscribeStreaming = @"transcribe";
```
[the remark this answers is not in the document]
Comment (palpatim) [reply AWSCore/Service/AWSService.m]: No, "transcribe" is the service name used for SigV4 credential scope. https://docs.aws.amazon.com/transcribe/latest/dg/websocket.html

File: AWSCore/Authentication/AWSSignature.m [hunk 2]
```diff
@@ -577,20 +646,25 @@ + (NSString *)getCanonicalizedQueryString:(NSString *)query {
...
+    NSCharacterSet *whitespaceChars = [NSCharacterSet whitespaceCharacterSet];
+
     NSMutableArray *sortedHeaders = [[NSMutableArray alloc] initWithArray:[headers allKeys]];
 
     [sortedHeaders sortUsingSelector:@selector(caseInsensitiveCompare:)];
 
     NSMutableString *headerString = [NSMutableString new];
     for (NSString *header in sortedHeaders) {
+        NSString *value = [headers valueForKey:header];
+        if (value != nil) {
```
[the remark this answers is not in the document]
Comment (palpatim) [reply AWSCore/Authentication/AWSSignature.m]: Swift habits :)

File: AWSTranscribeStreaming/Internal/AWSSRWebSocketDelegateAdaptor.m [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+
+#import <AWSCore/AWSCore.h>
+#import "AWSSRWebSocketDelegateAdaptor.h"
+#import "AWSTranscribeStreamingEventDecoder.h"
+#import "AWSTranscribeStreamingClientDelegate.h"
+
+@interface AWSSRWebSocketDelegateAdaptor ()
+
+@property (nonatomic, weak, readwrite) id<AWSTranscribeStreamingClientDelegate> clientDelegate;
+@property (nonatomic, readwrite) dispatch_queue_t callbackQueue;
```
[the remark this answers is not in the document]
Comment (palpatim) [reply AWSTranscribeStreaming/Internal/AWSSRWebSocketDelegateAdaptor.m]: Done

File: AWSTranscribeStreaming/AWSTranscribeStreamingClientDelegate.h [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    AWSTranscribeStreamingClientErrorCodeInvalidMessagePrelude,
+    AWSTranscribeStreamingClientErrorCodeInvalidMessageLengthHeader,
+    AWSTranscribeStreamingClientErrorCodeWebSocketProtocolError,
+    AWSTranscribeStreamingClientErrorCodeWebSocketCouldNotInitialize,
+    AWSTranscribeStreamingClientErrorCodeWebSocketClosedUnexpectedly,
+    AWSTranscribeStreamingClientErrorCodeUnknown
+};
+
+typedef NS_ENUM(NSInteger, AWSTranscribeStreamingClientDelegateConnectionStatus) {
+    AWSTranscribeStreamingClientDelegateConnectionStatusConnecting,
```
[the remark this answers is not in the document]
Comment (palpatim) [reply AWSTranscribeStreaming/AWSTranscribeStreamingClientDelegate.h]: Good call, done.

File: AWSTranscribeStreaming/Internal/AWSTranscribeStreamingChunkedEncodingInputStream.h [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+/**
+ * A subclass of NSInputStream that wraps an input stream and adds
+ * signature of chunk data.
+ *
+ * This class is created to handle `NSInputStream` as the input type
+ * One way of handling streaming is that we allow user to specify an input stream
+ * where they can keep writing data. We can then use that input stream to add headers,
+ * sign the request and publish the data packet in our outbound http2 stream
+ **/
+@interface AWSTranscribeStreamingChunkedEncodingInputStream : NSInputStream <NSStreamDelegate>
```
[the remark this answers is not in the document]
Comment (palpatim) [reply AWSTranscribeStreaming/Internal/AWSTranscribeStreamingChunkedEncodingInputStream.h]: Whoops, missed that during cleanup, thanks.

File: AWSTranscribeStreaming/AWSTranscribeStreamingService.m [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+        return self.webSocket;
+        
+    }
+}
+
+- (AWSTask *)getWebsocketForRequest:(NSDictionary *)requestParams {
+
+    return [[self getPreSignedURL:requestParams] continueWithBlock:^id _Nullable(AWSTask<NSURL *> * _Nonnull task) {
+        if (task.error != nil) {
+            return [AWSTask taskWithError:task.error];
```
[the remark this answers is not in the document]
Comment (palpatim) [reply AWSTranscribeStreaming/AWSTranscribeStreamingService.m]: Added logging in the parent method

File: AWSTranscribeStreaming/AWSTranscribeStreamingService.m [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+        networkingRequest.HTTPMethod = HTTPMethod;
+        networkingRequest.requestSerializer = [[AWSJSONRequestSerializer alloc] initWithJSONDefinition:[[AWSTranscribeStreamingResources sharedInstance] JSONObject]
+                                                                                            actionName:operationName];
+        networkingRequest.responseSerializer = [[AWSTranscribeStreamingResponseSerializer alloc] initWithJSONDefinition:[[AWSTranscribeStreamingResources sharedInstance] JSONObject]
+                                                                                                             actionName:operationName
+                                                                                                            outputClass:outputClass];
+
+        [[[self getWebsocketForRequest:networkingRequest.parameters] continueWithBlock:^id _Nullable(AWSTask * _Nonnull t) {
+            return nil;
+        }] waitUntilFinished];
```
[the remark this answers is not in the document]
Comment (palpatim) [reply AWSTranscribeStreaming/AWSTranscribeStreamingService.m]: Great callout! Handed the work off to the global high priority queue. I'm not able to think of a good way to test that there won't be a deadlock, but all existing tests work as expected even with the work being performed on global queue.

File: AWSCore/Authentication/AWSSignature.m [hunk 8]
```diff
@@ -370,77 +371,102 @@ - (NSString *)signRequestV4:(NSMutableURLRequest *)request
...
-        //Format of X-Amz-Credential : <your-access-key-id>/<date>/<AWS-region>/<AWS-service>/aws4_request.
-        NSString *scope = [NSString stringWithFormat:@"%@/%@/%@/%@",
-                           [currentDate aws_stringValue:AWSDateShortDateFormat1],
-                           endpoint.regionName,
-                           endpoint.serviceName,
-                           AWSSignatureV4Terminator];
-        
-        NSString *signingCredentials = [NSString stringWithFormat:@"%@/%@",credentials.accessKey, scope];
-        //need to replace "/" with "%2F"
-        NSString *xAmzCredentialString = [signingCredentials stringByReplacingOccurrencesOfString:@"/" withString:@"\%2F"];
```
Comment (royjit) [inline AWSCore/Authentication/AWSSignature.m]: This line is removed in the change, will this be handled by NSURLQueryItem?

Comment (palpatim) [reply AWSCore/Authentication/AWSSignature.m]: Yes, percent encoding is done at the point of conversion into a canonical request

**Pull request merged**